ethereum/contracts/bridge/L1WethBridge.sol

```diff
@@ -0,0 +1,369 @@
+        address _l1Sender,
+        address _l2Receiver,
+        uint256 _amount
+    ) internal pure returns (bytes memory txCalldata) {
+        txCalldata = abi.encodeCall(
+            IL2WethBridge.finalizeDeposit,
+            (_l1Sender, _l2Receiver, _amount)
+        );
+    }
+
+    /// @notice Withdraw funds from the initiated deposit, that failed when finalizing on L2.
+    /// Note: Refund is performed by sending equivalent amount of ETH to refund recipient address on L2.
+    function claimFailedDeposit(
+        address, // _depositSender,
+        bytes32, // _l2TxHash
+        uint256, // _l2BlockNumber,
+        uint256, // _l2MessageIndex,
+        uint16, // _l2TxNumberInBlock,
+        bytes32[] calldata // _merkleProof
+    ) external nonReentrant {
+    // ) external nonReentrant senderCanCallFunction(allowList) {
+        revert(
+            "Method not supported. ETH refund is handled by the zkSync contract."
+        );
+    }
+
+    /// @notice Finalize the WETH withdrawal and release funds
+    /// @param _l2BlockNumber The L2 block number where the WETH withdrawal was processed
+    /// @param _l2MessageIndexes The position in the L2 logs Merkle tree of the l2Logs that were sent with the ETH and WETH withdrawal messages
+    /// @param _l2TxNumberInBlock The L2 transaction number in a block, in which the ETH and WETH withdrawal logs were sent
+    /// @param _messages The L2 ETH and WETH withdraw data, stored in an L2 -> L1 messages
+    /// @param _merkleProofs The Merkle proofs of the inclusion L2 -> L1 messages about ETH and WETH withdrawal initializations
+    function finalizeWithdrawal(
+        uint256 _l2BlockNumber,
+        FinalizeWithdrawalL2MessageIndexes calldata _l2MessageIndexes,
+        uint16 _l2TxNumberInBlock,
+        FinalizeWithdrawalMessages calldata _messages,
+        FinalizeWithdrawalMerkleProofs calldata _merkleProofs
+    ) external nonReentrant {
+    // ) external nonReentrant senderCanCallFunction(allowList) {
+        require(!isWithdrawalFinalized[_l2BlockNumber][_l2MessageIndexes.wethL2MessageIndex], "WETH withdrawal is already finalized");
+
+        _proveL2MessagesInclusion(
+            _l2BlockNumber,
+            _l2MessageIndexes,
+            _l2TxNumberInBlock,
+            _messages,
+            _merkleProofs
+        );
+
+        (address _l1EthWithdrawReceiver, uint256 _ethAmount) = _parseL2EthWithdrawalMessage(_messages.ethMessage);
+        require(_l1EthWithdrawReceiver == address(this), "Wrong L1 ETH withdraw receiver");
+
+        (address _l1WethWithdrawReceiver, uint256 _wethAmount) = _parseL2WethWithdrawalMessage(_messages.wethMessage);
+        require(_l1WethWithdrawReceiver != address(0), "L1 WETH withdraw receiver can not be zero");
+
+        require(_ethAmount == _wethAmount, "Unequal ETH and WETH amounts in the L2 messages");
+
+        // Widthdraw ETH to smart contract address
+        zkSyncMailbox.finalizeEthWithdrawal(
```

Comment:
If ether is withdrawn before, `finalizeEthWithdrawal` Mailbox function will revert and the L1 bridge won't be able to collect the funds and send them to the L1 user.
Our L2 bridge contract checks if WETH message is already processed and Mailbox checks the same for ETH message.
Do you think we should handle something additional here?

Reply:
Imagine Alice initialize the withdrawal from L2. Then Alice wants to get funds on L1, so she just calls `finalizeWithdrawal` on `L1WethBridge` contract. 

But Bob see that Alice wants to `finalizeWithdrawal`, he takes the Merkle proof of `finalizeEthWithdrawal` and fronton Alice but finislizing the eth withdrawal directly to the zkSync contract. Now, the ether is withdrawn to the L1WethBridge contract. 

Alices transaction fails since the `finalizeEthWithdrawal` fails (it is already done and can't be done one more time).

Reply:
Good catch, makes sense